### PR TITLE
Fix infinite recursion bug causing “Maximum call stack size exceeded” error

### PR DESCRIPTION
**Pull request by samhocevar**
When calling _e.g._ `npm install`, `failedDependency()` has a potential unchecked infinite recursive call, causing `npm` to crash in some situations, when circular dependencies are involved.

I believe this PR fixes the bug, let me know of anything I can do to improve it if necessary.

## References

As a Google search indicates, dozens if not hundreds of users have reported issues related to this overflow, it even [appears to cause BSODs](https://github.com/microsoft/WSL/issues/436). The underlying bug has never been really investigated and some of these issues have been closed due to inactivity, or due to users clearing their `npm` cache, or managing to install packages through a slightly different invocation.

 - https://github.com/facebook/create-react-app/issues/7674
 - https://github.com/frctl/fractal/issues/487
 - https://github.com/nodejs/node/issues/23446
 - https://github.com/npm/npm/issues/17056#issuecomment-402643258
 - https://github.com/facebook/create-react-app/issues/7264
 - https://github.com/angular/angular-cli/issues/12885
 - https://github.com/webpack/webpack/issues/8998
 - https://github.com/laravel/framework/issues/27562
 - https://github.com/cybercase/netmd-js/issues/3
 - https://github.com/vuejs/vue-cli/issues/2567
 - https://github.com/svanderburg/node2nix/issues/121
 - https://github.com/dcodeIO/long.js/issues/73
 - https://github.com/JedWatson/classnames/issues/203
 - https://github.com/ionic-team/ionic-framework/issues/21523
 - https://github.com/apexcharts/react-apexcharts/issues/204
 - https://github.com/webpack/webpack/issues/8795
 - https://github.com/angular/angular-cli/issues/12231
 - https://github.com/Varying-Vagrant-Vagrants/VVV/issues/2024
 - https://github.com/nwjs/npm-installer/issues/73
 - https://github.com/xtuc/webassemblyjs/issues/486
 - https://github.com/microsoft/WSL/issues/436
 - https://github.com/rgbkrk/atom-script/issues/1981
 - https://github.com/lerna/lerna/issues/1457#issuecomment-629985231
 - https://github.com/zkat/npx/issues/112#issuecomment-473678507

### CODE REVIEW

File: lib/install/deps.js [hunk 1]
```diff
@@ -451,10 +451,13 @@ function failedDependency (tree, name, pkg) {
...
   if (!tree.requiredBy) return false
 
+  doneList = doneList || {}
+  doneList[tree] = true
+
   let anyFailed = false
   for (var ii = 0; ii < tree.requiredBy.length; ++ii) {
     var requireParent = tree.requiredBy[ii]
-    if (failedDependency(requireParent, moduleName(tree), tree)) {
+    if (!doneList[requireParent] && failedDependency(requireParent, moduleName(tree), tree, done)) {
```
Comment (Fadoli) [inline lib/install/deps.js]: this should be

```if (!doneList[requireParent] && failedDependency(requireParent, moduleName(tree), tree, doneList)) {```

Comment (samhocevar) [reply lib/install/deps.js]: Thanks. I fixed this by using `done` everywhere, to match the coding style of `loadRequestedDeps()` or `getAllMetadata` in the same file.

**Review by lyrenhexfan1 — CHANGES_REQUESTED** [review]
i like it the way it was

File: lib/install/deps.js (hunk 1 above)
Comment (Fadoli) [reply lib/install/deps.js]: done is actually a keyword used for callback in most js program so it was better with the older variable name.

I'm not a maintainer of this project, and it doesn't seems like one of them is looking at this PR anyway

**Pull request closed without merging**